### PR TITLE
[inductor] Handle more edge cases in slice and slice_scatter

### LINKED ISSUE #117110: [inductor][cpu]pyhpc_isoneutral_mixing meet LoweringException: AssertionError:
### 🐛 Describe the bug

pyhpc_isoneutral_mixing meet LoweringException: AssertionError: 
```
loading model: 0it [00:00, ?it/s]cpu  eval  pyhpc_isoneutral_mixing            

ERROR:common:backend='inductor' raised:
LoweringException: AssertionError: 
  target: aten.slice_scatter.default
  args[0]: TensorBox(
    ReinterpretView(
      StorageBox(
        InputBuffer(name='arg19_1', layout=FixedLayout('cpu', torch.float64, size=[2, 2, 1, 2, 2], stride=[8, 4, 4, 2, 1]))
      ),
      FixedLayout('cpu', torch.float64, size=[0, 2, 1, 2, 2], stride=[8, 4, 4, 2, 1]),
      origins={slice_229}
    )
  )
  args[1]: TensorBox(StorageBox(
    Pointwise(
      'cpu',
      torch.float64,
      def inner_fn(index):
          i0, i1, _, i3, i4 = index
          tmp0 = ops.index_expr(0, torch.int64)
          tmp1 = ops.index_expr(1, torch.int64)
          tmp2 = tmp0 >= tmp1
          tmp3 = ops.constant(nan, torch.float64)
          tmp4 = ops.masked(tmp2, tmp3, 0.0)
          tmp5 = ops.load(arg19_1, i4 + 2 * i3 + 4 * i1 + 8 * i0)
          tmp6 = ops.where(tmp2, tmp4, tmp5)
          return tmp6
      ,
      ranges=[0, 0, 1, 2, 2],
      origin_node=slice_scatter_33,
      origins={add_5, select_scatter_1, minimum, sub_9, tanh, copy_...
    )
  ))
  args[2]: 1
  args[3]: 2
  args[4]: -2

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information

```

### Versions

Repro
[inductor_single_run.sh](https://github.com/chuanqi129/inductor-tools/blob/main/scripts/modelbench/inductor_single_run.sh)
bash inductor_single_run.sh single inference accuracy torchbench pyhpc_isoneutral_mixing float32 first static default 0
Suspected guilty commit: https://github.com/pytorch/pytorch/commit/f6be25bae60853c6654d0efea096a459afa72c01

</table><p>Reference SW info(nightly)</p><table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>item</th>
      <th>commit</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>torchbench</td>
      <td>7617d3f5</td>
    </tr>
    <tr>
      <td>torch</td>
      <td>2.3.0a0+git7d163a9</td>
    </tr>
    <tr>
      <td>torchvision</td>
      <td>0.16.0a0+8636bf3</td>
    </tr>
    <tr>
      <td>torchtext</td>
      <td>0.16.0a0+142d029</td>
    </tr>
    <tr>
      <td>torchaudio</td>
      <td>2.2.0a0+475b6ae</td>
    </tr>
    <tr>
      <td>torchdata</td>
      <td>0.7.0a0+eb9bf61</td>
    </tr>
    <tr>
      <td>dynamo_benchmarks</td>
      <td>e6efc29</td>
    </tr>
  </tbody>

[torchbench-pyhpc_isoneutral_mixing-inference-amp-static-default-accuracy-single-crash_guilty_commit.log](https://github.com/pytorch/pytorch/files/13888737/torchbench-pyhpc_isoneutral_mixing-inference-amp-static-default-accuracy-single-crash_guilty_commit.log)
cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @WeizhuoZhang-intel @chuanqi129

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117377

Fixes #117110

When slicing we can end up with start and end which are out of bounds, which is
handled in python slicing by clamping to the correct bounds. There is also the
case where end < start which should result in an empty slice.

In the isoneutral_mixing failure we have the second case, with `start=2, end=0`
which in `slice_scatter` became `src_size[dim] = -2`.

This PR improves slice's edge case handling and factors the start and end
normalization code out so it can be shared with slice_scatter.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -7772,15 +7809,6 @@ def fn(input_ids) -> torch.Tensor:
         x = torch.randn(2, 2)
         self.common(fn, (x,), atol=0, rtol=0)
 
-    # It's a view so it doens't generate a kernel
-    @expectedFailureCodegenDynamic
-    def test_slice(self):
```
Comment (peterbell10) [inline test/inductor/test_torchinductor.py:7777]: Unrelated change, I've moved this next to the other `test_slice` functions and renamed it `test_slice3` as there are already `test_slice` 1 and 2.

**Review by lezcano — APPROVED** [review]
